toucan_connectors/snowflake/snowflake_connector.py

```diff
@@ -282,174 +306,134 @@ def _refresh_oauth_token(self):
-        return connection
-
     def _set_warehouse(self, data_source: SnowflakeDataSource):
-        warehouse = data_source.warehouse
-        if self.default_warehouse and not warehouse:
-            data_source.warehouse = self.default_warehouse
-        return data_source
-
-    def _get_warehouses(self, warehouse_name: Optional[str] = None) -> List[str]:
-        with self._get_connection(warehouse=warehouse_name) as connection:
-            result = SnowflakeCommon().get_warehouses(connection, warehouse_name)
-        return result
+        return (
+            data_source.copy(update={'warehouse': self.default_warehouse})
+            if not data_source.warehouse
+            else data_source
+        )
 
-    def _get_databases(self, database_name: Optional[str] = None) -> List[str]:
-        with self._get_connection(database=database_name) as connection:
-            result = SnowflakeCommon().get_databases(connection, database_name)
-        return result
+    @overload
+    def _execute_query(
+        self,
+        query: str,
+        parameters: dict | list[str] | None = None,
+        *,
+        warehouse: str | None = None,
+        database: str | None = None,
+        as_df: Literal[True] = ...,
+    ) -> pd.DataFrame:
+        ...
 
-    def _retrieve_data(self, data_source: SnowflakeDataSource) -> pd.DataFrame:
-        with self._get_connection(data_source.database, data_source.warehouse) as connection:
-            result = SnowflakeCommon().retrieve_data(connection, data_source)
-        return result
+    @overload
+    def _execute_query(
+        self,
+        query: str,
+        parameters: dict | list[str] | None = None,
+        *,
+        warehouse: str | None = None,
+        database: str | None = None,
+        as_df: Literal[False],
+    ) -> list[dict]:
+        ...
+
+    def _execute_query(
+        self,
+        query: str,
+        parameters: dict | list[str] | None = None,
+        *,
+        warehouse: str | None = None,
+        database: str | None = None,
+        as_df: bool = True,
+    ) -> pd.DataFrame | list[dict]:
+        with self._get_connection(database=database, warehouse=warehouse) as conn:
+            curs = conn.cursor(SfDictCursor)
+            query_result = curs.execute(query, parameters)
+            assert query_result is not None
+            # snowflake typing is incomplete for DictCursor
+            results: list[dict] = query_result.fetchall()  # type:ignore[assignment]
+            return pd.DataFrame(results) if as_df else results
+
+    def _describe_query(self, query: str) -> dict[str, str]:
+        with self._get_connection() as conn:
+            curs = conn.cursor(SfDictCursor)
+            return {r.name: _TYPE_CODE_MAPPING[r.type_code] for r in curs.describe(query)}
+
+    def _get_warehouses(self, warehouse_name: str | None = None) -> list[str]:
+        query = 'SHOW WAREHOUSES'
+        if warehouse_name:
+            query += f"LIKE '{warehouse_name}'"
```

Comment:
there's a space missing, this will give `WAREHOUSESLIKE`

Reply:
Woops good catch! Anyway, I want to test this with a real snowflake deployement before merging, since we're accessing the `"name"` column of a DF, and Snowflake might return `NAME`.